File: src/DynamicProgramming/122-BestTimetoBuyandSellStockII/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution122
{
public:
  int maxProfit(vector<int> &prices)
  {
    int n = prices.size();
    vector<vector<int>> dp(n, vector<int>(2));

    // boundary condition
    // 有第一天 不持有股票和买入股票 两种状态
    dp[0][0] = 0;
    dp[0][1] = -prices[0]; // 买入即扣了这么多现金

    // 开始 dp，因为第一天初始情况以及考虑了，故我们从第二天开始
    for (int i = 1; i < n; i++)
    {
      // 即 昨天不持有股票延续到今天以及昨天持有股票但是今天卖了 这两种状态，选择收益大的那个
      dp[i][0] = max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
      // 昨天持有股票今天也没卖掉以及昨天没持有股票但是今天买入了 两种状态
      dp[i][1] = max(dp[i - 1][1], dp[i - 1][0] - prices[i]);
    }

    // 最终结果一定是最后一天没有买入股票的情况（不存在负值）
    return dp[n - 1][0];
  }
};

int main(void)
{

  return 0;
}

File: src/DynamicProgramming/122-BestTimetoBuyandSellStockII/main.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        @cache
        def dfs(i, hold):
            # 到了第一天如果选择买入就减第一天的股票钱，否则就不干任何事
            if i < 0:
                return -prices[0] if hold else 0

            # 这一天是 持有股票 状态
            # 只能从 昨天也是持有股票而且昨天今天什么都不敢 和 昨天没有持有股票但到了今天买入了股票 两个状态转移过来
            if hold:
                return max(dfs(i - 1, 1), dfs(i - 1, 0) - prices[i])
            # 这一天是 没有持有股票 状态
            # 只能从 昨天没有持有股票 和 昨天持有股票但今天卖了股票 得来
            return max(dfs(i - 1, 0), dfs(i - 1, 1) + prices[i])

        return dfs(len(prices) - 1, 0)
